README.md
malscan
============

ClamAV-based malware scanner for Linux web servers.

[![Latest version](https://img.shields.io/github/release/malscan/malscan.svg)](https://github.com/malscan/malscan/releases)
[![GitHub license](https://img.shields.io/github/license/malscan/malscan.svg)](https://github.com/malscan/malscan/blob/1.x/LICENSE)
[![Build status](https://gitlab.com/malscan/malscan/badges/1.x/pipeline.svg)](https://gitlab.com/malscan/malscan/pipelines)

# Table of Contents
* [Features](#features)
* [Requirements](#requirements)
* [Installation](#installation)
* [Changelog](#changelog)

## Summary

malscan is a scanning platform for Linux servers that simplifies keeping your
web servers secure and malware-free. It is built upon the ClamAV platform,
providing all of the features of Clamscan with a host of new features and
detection modes.

## Features
* Multiple channels of malware signatures
    * RFX Networks Signatures
    * Metasploit Signatures
    * malscan Signatures
    * ClamAV Main Signatures
* Multiple Detection Methods
    * Standard HEX or MD5 based detections
        * Includes a database of over 30,000 signatures
        * Uses both HEX and hash based detections
    * String length detections
        * Identifies long obfuscated strings.
        * Non-signature based, to help detect zero day code injections
    * MimeType mismatch detections
        * Detects PHP files that are masquerading as other file types.
        * Identifies certain common obfuscated attack vectors, such as command
        shells hiding as `.png` files.
    * Tripwire detection mode
        * Allows you to whitelist files in a known clean configuration.
        * Compares the file tree to previously whitelisted states to identify changes.
        * Can be hooked into common deploy tools, such as `capistrano` or `dpl`.
* Easy File Quarantining
* Custom email notifications

## Requirements
* Linux Server / Desktop
* Full root access, or the ability to install:
  * ClamAV
  * File
  * Postfix or Exim, if using email notifications.
* SSH Access

## Installation

__NOTE__: Malscan v1.x is the last to support RPM-based installation. Malscan v2.x will
support platform-agnostic installation processes, including Puppet and script-based installers.

#### Red Hat Enterprise Linux, CentOS, and Fedora

**NOTE**: Malscan RPM packages are depreciated and will be removed

RPM packages are available for RHEL, CentOS, and Fedora. Packages are generated
only for currently supported operating systems. Support includes:

| Operating System | Version |           |
| :--------------: | :-----: | :-------: |
|   RHEL / CentOS  |    7    | ![](https://img.shields.io/badge/supported-yes-blue.svg) |
|   RHEL / CentOS  |    6    | ![](https://img.shields.io/badge/supported-yes-blue.svg) |
|      Fedora      |    28   | ![](https://img.shields.io/badge/supported-yes-blue.svg) |
|      Fedora      |    27   | ![](https://img.shields.io/badge/supported-yes-blue.svg) |
|      Fedora      |    26   | ![](https://img.shields.io/badge/supported-no-red.svg) |

Guides on how to set up the malscan yum repository can be found here:

* [CentOS 7](https://malscan.readthedocs.io/en/latest/guides/installation/centos7/)
* [CentOS 6](https://malscan.readthedocs.io/en/latest/guides/installation/centos6/)

#### Other Operating Systems

We make every effort to support as many operating systems as possible. Currently, we
officially support the following operating systems:

* Alpine Linux
* Debian
* Unbuntu (LTS only)

All of the above are supported via the malscan installer. This installer will automatically
identify your operating system, and install required dependencies.

**NOTE:** For Alpine Linux, you *must* install bash before you can run the installer.

To install via the installer on a supported system, simply run:

``` bash
curl -sSL https://raw.githubusercontent.com/malscan/malscan/1.x/installer | bash
```

We strongly recommend reading through the installer before running it. Running scripts
without knowing what they do is dumb.

## Usage

See `malscan -h` for more detailed program usage.

## Contributing

If you're interested in contributing to malscan, I am looking for the following help:

* Feature development
* Documentation
* Web design
* Logo design

Contact me at `jgrancell@malscan.com` if you're interested.

## Licensing and Terms of Usage

The malscan application is released under the MIT license, which is included in the repository. This software is provided as is, with absolutely no warranty provided for it. As this application does alter file/directory structure for this server, make sure that you understand exactly what this application does and the security ramifications of it.


CHANGELOG.md
## malscan

#### Version 1.8.2
*Release: Jan 17, 2020*

* **IMPORTANT**: This will be the last non-maintenance release of the 1.x branch of Malscan. No
additional features will be coming to the 1.x branch.
* Fixed: Corrected all URLs within repository code and documentation post server move.
* Updated: Copyright now extends to 2020.

#### Version 1.8.1
*Release: Nov 26, 2018*
* Fixed: Malscan now provides better information when attempting a run with the lockfile present.

#### Version 1.8.0
*Release: Oct 15, 2018*
* New: Added a proper `CHANGELOG.md` file.
* Updated: Substantial Installer update
    * Fixed: Installer was referencing bad download URLs. Fixed.
    * Fixed: `which` may not be available in some distributions. Changed to `command -v`
    * Updated: Added support for Ubuntu 16.04 and 18.04.
    * Updated: Added support for Debian 8 and 9.
    * Updated: Added support for Alpine Linux.
    * Updated: Substantially refactored Installer code to remove duplicate lines.
* Updated: Fixed some non-portable code in the mimescan function.
* Updated: Cleaned up some shellcheck notices in the avscan function.

#### Version 1.7.2-1
*Release: May 15, 2018*
* Fixed: Updater will now properly pull the malscan core version from the right git branch.
* Updated: Updated all documentation to point to new docs site and new 1.x branch

#### Version 1.7.1
*Release: April 20, 2018*
* New: Tripwire detection mode
* Fixed: Corrected permission issue on log file. (fixes #15)
* Updated: Changed the save path for the Tripwire whitelist file

#### Version 1.7.0
*Release: March 22, 2018*
* Feature: Lock files are used to ensure multiple runs of malscan don't stack.
* Feature: Configuration options can now be viewed using the malscan -c command. (fixes #10)
* Feature: Configuration options can now be set using the malscan -s OPTION value command.
* Fixed: malscan will now correctly check for sudo.
* Fixed: malscan will now check to see if the user is in the malscan group, in lieu of being run as sudo.
* Fixed: Notifications are now sent in a more spam-checker-friendly format, reducing issues with notifications ending up in the spam folder.
* Updated: malscan will now use its own freshclam.conf file and /var/lib/malscan signatures directory, to prevent conflicts with ClamAV.
* Updated: The malscan file structure has been updated to conform with the FHS. (fixes #7)
* Updated: Rewrote the install.sh script to support Fedora, Debian, and CentOS/RHEL 7.
* Updated: Created RPM packaging for CentOS/RHEL 6, 7, and Fedora 26/27. (fixes #8)
* Updated: New exhaustive build testing CI pipeline for automated malscan testing.
* Removed: Removed whitelisting and tripwire scanning until it can be re-worked in a later release.
* Removed: Removed reporting until it can be re-developed in a later release.
* Removed: Removed Ubuntu/Debian support while working on packaging.

#### Version 1.5.3
*Release: September 4, 2015*
* Bugfix: Corrected text coloring on the update.sh script to terminate properly
* Bugfix: Removed some excess text from the Mimescan
* Updated: All malscan runs now include the current Malscan version and the time that signatures were last updated.
* Updated: Unified logging files for all scantypes into a single scan log for each scan
* Feature: Added new -u update functionality, which updates both the core application as well as the signatures

#### Version 1.5.2
*Release: July 13, 2015*
* Bugfix: Corrected the Mimetype scan to properly ignore files listed in conf.malscan

#### Version 1.5.1
*Release: June 16, 2015*
* Bugfix: Corrected a bug with update.sh causing a fatal error.
* Updated: Added output identifying when different scan types are starting, for more verbose and informative output.
* Updated: Incremented the version in malscan.sh
* Updated: Removed the version information in the comment header in install.sh, added a Since version header instead.
* Updated: Removed the output text from the freshclam updater, which was getting too messy.
* Updated: Added a warning in update.sh indicating the now-silent freshclam update portion can take a long time.

#### Version 1.5.0
*Release: June 1, 2015*
* Feature: Added automated whitelisting of file trees for known clean files (such as imports from development enviroments or fresh installs)
* Feature: New Tripwire scanning mode. Identifies any files that have been changed or did not exist from the whitelist reference. Excellent for static sites or minimally changing applications.
* Feature: New installer.
  * New installer is compatible with and fully tested on CentOS 6, CentOS 7, and Ubuntu 14.04.
    * Installer may work with other 6.x and 7.x RHEL derivatives as well as Ubuntu 12.x, 13.x, 14.x, and 15.x, however these are officially unsupported.
    * If you run into issues with any non-supported Operating Systems, please submit a Github issue so that I can correct it and add that OS to the supported list.
    * Added package installation sanity checking, to ensure everything is set up properly
  * New installer may work with other RHEL derivatives as well as RHEL/CentOS 5
* Updated: Mimetype scanning to add additional filetypes to the scan.
* Updated: A substantial number of prompts, both in text and color.

#### Version 1.4.4
*Released: May 6, 2015*
* Bugfix: Corrected an issue with notifications not being sent because there was no way to specify receiving email addresses. Fixed in conf.malscan-blank.
* Bugfix: Corrected an issue with whitelisting not working properly. It should now function correctly, and is working in test RHEL 6 and CentOS 7 testing environments.
* Special Note: The changes to conf.malscan-blank will need to be manually added to any active conf.malscan files.

#### Version 1.4.3
*Released: May 5, 2015*
* Bugfix: Corrected a logging path issue. All log files will now be correctly generated in the 'log' directory inside your chosen path in conf.malscan
* Bugfix: Corrected the URL for the custom virus definitions

install.sh
#!/bin/bash
# -------------------------------------------------
#
# Package: Malscan
# Author: Josh Grancell <josh@joshgrancell.com>
# Description: Linux malware scanner for web servers
# Copyright: 2015-2018 Josh Grancell
# License: MIT License
#
# -------------------------------------------------

MALSCAN_REPOSITORY_URL="https://gitlab.com/malscan/malscan/raw"
SIGNATURE_REPOSITORY_URL="https://www.joshgrancell.com/signatures"

## Determining which release to install
if [[ $# == 1 ]]; then
  CURRENT_INSTALLER_BRANCH="$1"
else
  CURRENT_INSTALLER_BRANCH="master"
fi

## Verifying that we are running under root/sudo
if [[ "$EUID" != 0 ]]; then
  echo -e "\033[31mThe installer must be run either as the root user or under sudo.\033[37m"
  exit 1
fi

## Determining our OS
if [[ -f "/etc/redhat-release" ]]; then
  ## Determining the RedHat-derivative OS
  if [[ -d /usr/local/cpanel ]]; then
    DISTRO="cPanel"
  elif [[ -f /etc/centos-release ]]; then
    DISTRO="CentOS"
  elif [[ -f /etc/fedora-release ]]; then
    DISTRO="Fedora"
  else
    DISTRO="RHEL"
    ## EPEL is required for installation
    EPEL_REQUIRED=1
  fi

elif [[ -f /etc/lsb-release ]]; then
  DISTRO=$(grep "DISTRIB_ID" /etc/lsb-release | cut -d= -f2)
  VERSION=$(grep "DISTRIB_RELEASE" /etc/lsb-release | cut -d= -f2)
else
  DISTRO="Unsupported"
fi

if [[ "$DISTRO" == "cPanel" ]]; then
  echo "Because this is a cPanel system, ClamAV must be manually installed through WHM."
  echo "To install ClamAV through WHM, use the cPanel > Manage Plugins tool."
  echo "If you have completed the ClamAV plugin installation, you may continue with the Malscan installation."
  echo "Continue? [y/N]"
  read INSTALL_OPTION

  if [[ "$INSTALL_OPTION" != "y" && "$INSTALL_OPTION" != "Y" ]]; then
    echo -e "\033[31mYou have selected to exit this isntallation. Please run this installer again once the ClamAV plugin has been installed.\033[37m"
    exit 0
  fi
  yum -y install file wget

elif [[ "$DISTRO" == "RHEL" || "$DISTRO" == "CentOS" ]]; then
  yum -y install epel-release file wget

  if grep -qs "release 6" /etc/redhat-release; then
    yum -y install clamav clamav-db
  else
    yum -y install clamav clamav-update
  fi
elif [[ "$DISTRO" == "Fedora" ]]; then
  ## Installing for Fedora
  dnf -y install clamav-clamav-update file weget
elif [[ "$DISTRO" == "Ubuntu" ]]; then
  ## Installing for Ubuntu
  apt-get update
  apt-get -y install clamav clamav-freshclam file wget
else
  ## Incompatible distro, exiting
  echo -e "\033[31mMalscan has detected that this server is not running a supported operating system."
  echo "Malscan is currently only available for installation on the following Operating Systems: "
  echo "    CentOS/RHEL 6.x / 7.x"
  echo "    Fedora 24 / 25 / 26"
  echo "    Ubuntu 14.04 / 16.04"
  echo -e "Feature requests for new Operating System support can be submitted to malscan.org/features\033[37m"
  exit 0
fi

# ----------------------------------------------------
# Malscan configuration section

getent group malscan > /dev/null || groupadd -r malscan
getent passwd malscan > /dev/null || useradd -r -g malscan -s /sbin/nologin -c "Malscan Service User" malscan

MALSCAN_BINARY_PATH="/usr/local/bin"
MALSCAN_SYSCONFIG_PATH="/etc/malscan"
MALSCAN_LOGGING_PATH="/var/log/malscan"
MALSCAN_SIGNATURE_PATH="/var/lib/malscan"
MALSCAN_MAN_PATH="/usr/local/share/man/man1"
MALSCAN_MISC_PATH="/usr/local/share/malscan"

## Setting up our configuration files
mkdir -p "$MALSCAN_SYSCONFIG_PATH"
wget -P "$MALSCAN_SYSCONFIG_PATH/" "$MALSCAN_REPOSITORY_URL/$CURRENT_INSTALLER_BRANCH/malscan.conf" --quiet
wget -P "$MALSCAN_SYSCONFIG_PATH/" "$MALSCAN_REPOSITORY_URL/$CURRENT_INSTALLER_BRANCH/freshclam.conf" --quiet
chown -R malscan:malscan "$MALSCAN_SYSCONFIG_PATH"

## Setting up our Signatures and Logging directories
mkdir -p "$MALSCAN_LOGGING_PATH"
chown -R malscan:malscan "$MALSCAN_LOGGING_PATH"
mkdir -p "$MALSCAN_SIGNATURE_PATH"
chown -R malscan:malscan "$MALSCAN_SIGNATURE_PATH"

## Setting up our man page
mkdir -p "$MALSCAN_MAN_PATH"
wget -P "$MALSCAN_MAN_PATH/" "$MALSCAN_REPOSITORY_URL/$CURRENT_INSTALLER_BRANCH/malscan.1" --quiet

## Setting up the malscan binary
mkdir -p "$MALSCAN_BINARY_PATH"
echo "$MALSCAN_REPOSITORY_URL/$CURRENT_INSTALLER_BRANCH/malscan"
wget -P "$MALSCAN_BINARY_PATH" "$MALSCAN_REPOSITORY_URL/$CURRENT_INSTALLER_BRANCH/malscan" --quiet
chmod +x "$MALSCAN_BINARY_PATH/malscan"

## Creating the misc file directory
mkdir -p "$MALSCAN_MISC_PATH"
wget -P "$MALSCAN_MISC_PATH" "$MALSCAN_REPOSITORY_URL/$CURRENT_INSTALLER_BRANCH/LICENSE" --quiet
wget -P "$MALSCAN_MISC_PATH" "$MALSCAN_REPOSITORY_URL/$CURRENT_INSTALLER_BRANCH/README.md" --quiet
wget -P "$MALSCAN_MISC_PATH" "$MALSCAN_REPOSITORY_URL/$CURRENT_INSTALLER_BRANCH/version.txt" --quiet
chown -R malscan:malscan "$MALSCAN_MISC_PATH"

## Getting the user's input on email notifications
echo -e "\033[33mBeginning the malscan configuration process..."
echo ""
echo -ne "Would you like to enable email notifications? [y/N] [default: N] \033[37m"
read EMAIL_NOTIFICATIONS

if [[ "$EMAIL_NOTIFICATIONS" == "y" || "$EMAIL_NOTIFICATIONS" == "Y" || "$EMAIL_NOTIFICATIONS" == "yes" || "$EMAIL_NOTIFICATIONS" == "YES" ]]; then
    ## Email notifications are being enabled, so we're now getting the list of addresses
    echo -e "\033[33mAt what email addresses would you like to receive notifications (Comma-separated list): \033[37m"
    read EMAIL_ADDRESSES

    echo -e "\033[33mWhat email address would you like malscan to send email from? \033[37m"
    read SENDER_ADDRESS

    "$MALSCAN_BINARY_PATH/malscan" -s EMAIL_NOTIFICATIONS TRUE > /dev/null
    "$MALSCAN_BINARY_PATH/malscan" -s NOTIFICATION_ADDRESS "$EMAIL_ADDRESSES" > /dev/null
    "$MALSCAN_BINARY_PATH/malscan" -s MALSCAN_SENDER_ADDRESS "$SENDER_ADDRESS" > /dev/null

fi

## No remote quarantine, so we're now requesting the local quarantine directory
echo ""
echo -e "\033[33mWhat directory would you like to quarantine files in? [default=/root/.malscan/quarantine] \033[37m"
read QUARANTINE_DIRECTORY

if [[ "$QUARANTINE_DIRECTORY" = "" || "$QUARANTINE_DIRECTORY" = " " ]]; then
    QUARANTINE_PATH="/root/.malscan/quarantine"
else
    QUARANTINE_PATH="$QUARANTINE_DIRECTORY"
fi

## Creating the quarantine path if it doesn't exist
if [[ -d "$QUARANTINE_PATH" ]]; then
    echo -e "\033[32mYour quarantine path has been successfully set!\033[37m"
    "$MALSCAN_BINARY_PATH/malscan" -s QUARANTINE_DIRECTORY "$QUARANTINE_PATH" > /dev/null
else
    mkdir -p "$QUARANTINE_PATH"
    echo -e "\033[32mThe directory $QUARANTINE_PATH has been created, and set as the Quarantine location.\033[37m"
    "$MALSCAN_BINARY_PATH/malscan" -s QUARANTINE_DIRECTORY "$QUARANTINE_PATH" > /dev/null
fi

echo -e "\033[032mMalscan has been successfully configured! Beginning initial update...\033[37m"

echo -e "Pre-seeding the malscan signature databases with bundled signatures. This may take some time to complete."
wget -P "$MALSCAN_SIGNATURE_PATH/" "$SIGNATURE_REPOSITORY_URL/main.cvd" --quiet
wget -P "$MALSCAN_SIGNATURE_PATH/" "$SIGNATURE_REPOSITORY_URL/bytecode.cvd" --quiet
wget -P "$MALSCAN_SIGNATURE_PATH/" "$SIGNATURE_REPOSITORY_URL/daily.cvd" --quiet
chown -R malscan:malscan "$MALSCAN_SIGNATURE_PATH"

echo -e "Updating to the latest malscan signature versions. You can always do this using the command 'malscan -u'"

"$MALSCAN_BINARY_PATH/malscan" -u


echo "Malscan has been successfully configured and instantited."
echo "The current malscan configuration can be seen using the 'malscan -c' command"
echo "All chosen configuration options can be changed using the 'malscan -s' command"
echo "For more information on using malscan, use the 'malscan -h' or 'malscan --help' command."
echo ""
